### ISSUE TITLE
Unexpected error "The current Merlin configuration has been generated by another, incompatible, version of Dune"

### ISSUE THREAD

**Issue by erikmd** [ocaml/merlin#1591]
Dear all,

I recently reinstalled merlin in a local opam switch for working on [learn-ocaml](https://github.com/ocaml-sf/learn-ocaml),
and I systematically got the following error:

```
The current Merlin configuration has been generated by another, incompatible, version of Dune.
Please rebuild the project. (Using the same version of Dune as the one running the `ocaml-merlin` server.)
(431 pending errors, use C-c C-x to jump)
```

Thus:

* either there is a true incompatibility and an opam constraint should be added in **opam-repository**
* or I'm doing something wrong.

## Additional info

* Here is the list of our dependencies: [learn-ocaml.opam](https://github.com/ocaml-sf/learn-ocaml/blob/master/learn-ocaml.opam), thus requiring OCaml 4.12, and allowing dune upto 2.9.3
* `opam show merlin.4.7-412` currently says `"dune" {>= "2.9.0"}`
* Doing `rm -fr _opam` or `rm -fr _build` and starting over is not enough.
* ~`opam install merlin.4.6-412` solves the issue.~ [actually not :frowning_face:]

Thanks for any feedback allowing to fix this!

**Comment by erikmd** [comment]
This is really a blocking, usability issue of the merlin plugin for Emacs.

FTR:

* I use (ocaml)merlin and dune from opam,  
  and the merlin mode from MELPA (let me know if you need more details on my `.emacs`).
* The message `The current Merlin configuration has been generated by another, incompatible, version of Dune.` appears rather often (maybe systematically with several (Merlin, Dune) versions, a priori compatible w.r.t. their opam packages.  
I guess one should investigate the exact combination of (ocamlmerlin, Dune) that triggers a systematic error.
* Here is the list of workarounds I found so far to try avoiding the error:
  * Compile learn-ocaml (requiring ocaml = 4.12) with dune = 2.9.1, merlin = 4.5-412, yojson = 1.7.0 < 2.0.0
  * Then, the error probability < 1.
  * In case of error: `find _build -name "*merlin*" -exec echo rm -fr "{}" \;`
  * eval $(opam env) && make && make opaminstall
  * In Emacs: `M-x merlin-stop-server RET`, then `C-x C-s`

Hoping there is a better way to avoid this error (that didn't occur last year with a similar config)

**Comment by erikmd** [comment]
BTW, could you point me to the LoC where the error message "The current Merlin configuration has been generated by another, incompatible, version of Dune. Please rebuild the project. (Using the same version of Dune as the one running the `ocaml-merlin` server.)" is defined?

I was unable to find it in this repo 😅

**Comment by voodoos** [comment]
I have not tried hacking on learn-ocaml recently but does not remind having any issues last time I did.

Merlin runs a Dune process to get it's configuration, this means that the Dune that built the project should be the same than the one ran by Merlin to get the configuration. You should start emacs from a terminal that is in the correct switch to ensure that the correct version of Dune is in path.

There might be an issue with how learn OCaml is configured. Could you give me precise steps for me to try and reproduce the issue ?

(the error message actually comes from Dune: https://github.com/ocaml/dune/blob/643b4a3e26a331ad1a58e3fada5f57083e49ca74/src/dune_rules/merlin/merlin.ml#L115

**Comment by erikmd** [comment]
Hi @voodoos, thanks a lot for your detailed reply! that helped much.

> Could you give me precise steps for me to try and reproduce the issue ?

Here is the summary of all the steps (even if I'm not sure you'd need to replay/investigate all this, given you pinpointed the crux of the issue, see below):

<details><summary>Summary</summary>

```sh
opam switch create 4.14.2 # create a "main switch", with a more recent ocaml/merlin than that of learn-ocaml
eval $(opam env)
opam install merlin -y # → merlin.4.8-414

mkdir -p ~/forge/git && cd ~/forge/git
git clone https://github.com/ocaml-sf/learn-ocaml.git
cd learn-ocaml
opam switch create . ocaml-base-compiler.4.12.1 --deps-only -y # create a local switch for learn-ocaml
eval $(opam env)
opam install merlin -y # → merlin.4.7-412
make
make opaminstall
# learn-ocaml build serve --repo=demo-repository # run the server astest
```

Regarding the emacs config, I use [this `.emacs`](https://github.com/erikmd/tapfa-init.el/blob/master/.emacs) file, along with the following snippet stored in `~/forge/git/learn-ocaml/.dir-locals.el`, with variables to force using (`ocaml`, `ocamlmerlin`) binaries from the local switch:
```elisp
((tuareg-mode
  . (
     (tuareg-opam-insinuate . nil)
     (tuareg-interactive-program "opam exec --switch=$HOME/forge/git/learn-ocaml -- ocaml")
     (merlin-command
      . (concat (s-trim-right (shell-command-to-string "opam var --switch=$HOME/forge/git/learn-ocaml bin")) "/ocamlmerlin"))
     )))
```

</details>

### TL;DR:

The crux of the issue comes exactly from what you said *the Dune that built the project should be the same than the one ran by Merlin to get the configuration. You should start emacs from a terminal that is in the correct switch.*

However in terms of usability, this situation is not really satisfactory, i.e. when the error shows, we need to follow these steps (not self-evident w.r.t. the message raised by dune):

* close emacs
* open a terminal
* cd in the project dir
* run `eval $(opam env)`
* run `emacs &`
* reopen the relevant .ml file
* run `M-x merlin-stop-server`
* save with `C-x C-s`
* and finally see in the mini-buffer `No errors` :)

So actually, addressing the issue amounts to the following problem: we need to ensure that the `ocamlmerlin` binary is run with the proper PATH w.r.t. `dune`, so that even if the emacs instance is run from another initial PATH or directly from the window manager, the local development environment (which can be easily tuned in a `.dir-locals.el` file is consistent.

IOW, do you think it's easy to `(setq` another emacs variable in `.dir-locals.el` so that `ocamlmerlin` is run this way w.r.t. `dune` every time? or does it need further changes in `merlin.el`?

**Comment by erikmd** [comment]
> IOW, do you think it's easy to (setq another emacs variable in .dir-locals.el so that ocamlmerlin is run this way w.r.t. dune every time? or does it need further changes in merlin.el?

On a related note, we had a similar usability issue in the [emacs mode for Coq](https://github.com/ProofGeneral/PG/) that I co-maintain (it was necessary to restart emacs for any switch change), and @hendriktews wrote a very nice minor mode: [`opam-switch-mode`](https://github.com/ProofGeneral/opam-switch-mode), that is language-agnostic, and allows one to hot-switch between several opam switches from a menu, without restarting emacs.

So, even if #1591 could be solved by adding a variable in `.dir-locals.el` (I hope so),
maybe this issue could be the starting point of adapting `tuareg` & `merlin` to provide `opam-switch-mode` support? ✨🙂✨
(*see also https://proofgeneral.github.io/doc/master/userman/Coq-Proof-General/#index-opam_002dswitch_002dmode-support for context/example*)

Cc @monnier FYI
